yoyodyne/models/hard_attention.py

```diff
@@ -0,0 +1,481 @@
+            predictions = torch.cat((predictions, pred), dim=-1)
+            # Updates likelihood emissions.
+            likelihood = likelihood + self._gather_at_idx(
+                log_probs, pred, self.pad_idx
+            )
+        return predictions, likelihood
+
+    @staticmethod
+    def greedy_step(log_probs, likelihood):
+        # Updates current likelihood with greedy probability prediction.
+        tgt_prob = likelihood + log_probs.transpose(1, 2)
+        tgt_prob = tgt_prob.logsumexp(dim=-1)
+        tgt_char = torch.max(tgt_prob, dim=-1)[1]
+        return tgt_char.unsqueeze(-1), likelihood
+
+    @staticmethod
+    def _gather_at_idx(
+        prob: torch.Tensor, tgt: torch.Tensor, pad_idx=None
+    ) -> torch.Tensor:
+        """Collects probability of tgt index across all states in prob.
+
+        To calculate the final emission probability, the pseudo hmm
+        graph needs to aggregate the final emission probabilities of
+        tgt char across all potential hidden states in prob.
+
+        Args:
+            prob (torch.Tensor): log probs of emission states of shape
+                (batch_size x src_len x vocab_size)
+            tgt (torch.Tensor): tgt symbol to poll probabilities for
+                shape (batch_size).
+
+        Returns:
+            (torch.Tensor) emission probabilities of tgt symbol for
+                each hidden state of size (batch_size 1, src_len).
+        """
+        batch_size, src_seq_len, _ = prob.shape
+        idx = tgt.view(-1, 1).expand(batch_size, src_seq_len).unsqueeze(-1)
+        output = torch.gather(prob, -1, idx).view(batch_size, 1, src_seq_len)
+        idx = idx.view(batch_size, 1, src_seq_len)
+        if pad_idx:
+            pad_mask = (idx != pad_idx).float()
+        return output * pad_mask
+
+    @staticmethod
+    def _apply_mono_mask(transition_prob: torch.Tensor) -> torch.Tensor:
+        """Applies monotonic attention mask to transition probabilities.
+
+        Enforces a 0 log-probability values for all non-monotonic relations
+        in the transition_prob tensor (i.e. all values i < j per row j).
+
+        Args:
+            transition_prob (torch.Tensor): transition probabilities
+                between all hidden states (source sequence) of shape
+                (batch_size x src_len x src_len).
+
+        Returns:
+            (torch.Tensor) of masked transition probabilities of shape
+                (batch_size x src_len x src_len).
+
+        """
+        mask = torch.ones_like(transition_prob[0]).triu().unsqueeze(0)
+        # 0 log-probability value for masking.
+        # Implemented same as in original repo.
+        mask = (mask - 1) * defaults.NEG_LOG_EPSILON
+        transition_prob = transition_prob + mask
+        transition_prob = transition_prob - transition_prob.logsumexp(
+            -1, keepdim=True
+        )
+        return transition_prob
+
+    def training_step(self, batch: data.PaddedBatch, batch_idx: int) -> Dict:
```

Comment:
Is this return type correct? Does it not return a tensor (scalar) loss (sorry if this error is just copied from an earlier error))

Reply:
no you're right. probably an artifact from when i was trying something else out.